### PR TITLE
Add TheGreatWar project that depends on the IsometricEngine project. …

### PR DESCRIPTION
**Pull request by danieldownes**
Applied formatting to some code.
Moved some game specific classes to the separated game dedicated project.

### CODE REVIEW

File: IsometricEngine/include/AgentManager.h [hunk 1]
```diff
@@ -1,41 +1,41 @@
 #pragma once
 #include <cmath>
+#include <functional>
```
Comment (danieldownes) [inline IsometricEngine/include/AgentManager.h:3]: Only formatting was applied to this class, it looks like it these are now alphabetically sorted.

File: IsometricEngine/include/AgentManager.h [hunk 2]
```diff
@@ -1,41 +1,41 @@
...
-		InputState& state,
-		std::set<std::vector<BattlefieldCell>::iterator>* gameScene,
-		GameStateManager& gameStateManager,
-		Camera& camera,
-		Scene& scene);
+   public:
+	void onUpdate(InputState& state, std::set<std::vector<BattlefieldCell>::iterator>* gameScene,
+				  GameStateManager& gameStateManager, Camera& camera, Scene& scene);
 
-	void placeScenery(sf::Vector2i isometricCell, std::set<std::vector<BattlefieldCell>::iterator>* gameScene, Scenery scenObject, GameStateManager& gameStateManager);
```
Comment (danieldownes) [inline IsometricEngine/include/AgentManager.h:30]: Ah, with the exception of this function, which was more game specific as it only served to place a tree, which is only called.

File: IsometricEngine/include/GameState.h [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#pragma once
```
Comment (danieldownes) [inline IsometricEngine/include/GameState.h:1]: Only formatting applied over the next 3 files.

File: IsometricEngine/include/QuadTree.h [hunk 4]
```diff
@@ -0,0 +1,192 @@
+#pragma once
+
+#include <array>
+#include <iostream>
+#include <vector>
+
+#include "GlobalConstants.h"
+#include "SFML/Graphics/Rect.hpp"
+
+/**
```
Comment (danieldownes) [inline IsometricEngine/include/QuadTree.h:10]: Formatting and doxygen style brief comment.

File: IsometricEngine/premake5.lua [hunk 5]
```diff
@@ -1,5 +1,5 @@
-project "TheGreatWar"
-   kind "ConsoleApp"
+project "IsometricEngine"
+   kind "StaticLib"
```
Comment (danieldownes) [inline IsometricEngine/premake5.lua:2]: We now build the IsometricEngine as a lib. The game becomes the consoleApp that uses this lib.

File: IsometricEngine/src/AgentManager.cpp [hunk 6]
```diff
@@ -56,25 +54,6 @@ void AgentManager::onUpdate(
     }
 }
 
-
-void AgentManager::placeScenery(sf::Vector2i isometricCell, std::set<std::vector<BattlefieldCell>::iterator>* gameScene, Scenery sceneObject, GameStateManager& gameStateManager)
```
Comment (danieldownes) [inline IsometricEngine/src/AgentManager.cpp:60]: Moved to game project as it expects "Tree". For now we can keep this out of the engine, but perhaps supporting some generic terrain objects is a good thing. I wasn't happy about the type check using a string, and it can check a class type if needed at all.

File: IsometricEngine/src/Scene.cpp [hunk 7]
```diff
@@ -4,48 +4,58 @@
 
 Scene::Scene() {}
 
-void Scene::UpdateGameScene(Camera& cam, GameState& gameState, InputState& inputState) {
+void Scene::UpdateGameScene(Camera& cam, GameState& gameState, InputState& inputState)
```
Comment (danieldownes) [inline IsometricEngine/src/Scene.cpp:7]: Only format changes in this file.

File: TheGreatWar/premake5.lua [hunk 8]
```diff
@@ -0,0 +1,40 @@
+project "TheGreatWar"
```
Comment (danieldownes) [inline TheGreatWar/premake5.lua:1]: The new project which *links* IsometricEngine, and is able to use it due to the *includedirs*.

File: TheGreatWar/src/TheGreatWar.cpp [hunk 9]
```diff
@@ -48,3 +48,23 @@ int main()
 
 	return 0;
 }
+
+
+void placeTree(sf::Vector2i isometricCell, std::set<std::vector<BattlefieldCell>::iterator>* gameScene,
```
Comment (danieldownes) [inline TheGreatWar/src/TheGreatWar.cpp:53]: This was origonally placeScenary, but moved here, as its only called from here and needs work to make it not depend on a specific scenary type.

File: premake5.lua [hunk 10]
```diff
@@ -1,6 +1,6 @@
 -- premake5.lua
 workspace "TheGreatWar"
```
Comment (danieldownes) [inline premake5.lua:2]: The solution now include TheGreatWar and makes it the default startup project. Interestingly I found VS may have some caching issue, if you've already opened the solution. So you may have to manually set the start project. I added an issue to Premake github issues for others to have a look, but I believe is a VS thing.

**Review by 123samueld — COMMENTED** [review]
You're going to have to walk me through how to set this up because right now everything is broken. This is potentially too many changes to quickly. Splitting up engine and game concerns is a great idea but it's potentially pushing our time/skill/energy resources too far.

**Comment by danieldownes** [comment]
@123samueld 

> You're going to have to walk me through how to set this up because right now everything is broken. This is potentially too many changes to quickly. Splitting up engine and game concerns is a great idea but it's potentially pushing our time/skill/energy resources too far.

It would be handy if you posted the error messages you are getting, is it during premake, build or runtime? etc..

**Comment by danieldownes** [comment]
I would say splitting up the code base is a good investment, it saves time by divide and conquer method (not as in the game).

**Comment by 123samueld** [comment]
> @123samueld
> 
> > You're going to have to walk me through how to set this up because right now everything is broken. This is potentially too many changes to quickly. Splitting up engine and game concerns is a great idea but it's potentially pushing our time/skill/energy resources too far.
> 
> It would be handy if you posted the error messages you are getting, is it during premake, build or runtime? etc..

Yes, sorry I should have been more specific. It's the entire structure, when I pull the branch, rather than seeing the IsometricEngine and TheGreatWar folders like I can see here, there's only TheGreatWar and it doesn't appear to have had any files separated but does seem to have lots of code missing, probably .h files.

So just getting the new structure setup doesn't seem to be working when pulling this branch to my local repo.

File: IsometricEngine/src/AgentManager.cpp (hunk 6 above)
Comment (123samueld) [reply IsometricEngine/src/AgentManager.cpp:60]: Changing "string" to object is a good idea. 

There's a distinction between "terrain" and "scenery". You could think of it as a table cloth with crockery & cutlery on it, the terrain is the underlying table cloth and has some interaction but not much, the crockery & cutlery is the scenery and has lots of interaction.

- Terrain is the underlying map grid, like the grass sprites, they will have some interactivity but not much, just with explosions really. It's more about visualisation of the map's structure.

- Scenery is a child of the Agent class and all agents are where the interactivity is really at. You'll see the class "mobileAgents", Scenery is the opposite ie stationary Agents. Things that can be interacted with like trees (burn, cut, blast), buildings (place and build) etc but are not mobile so don't need Boid (flocking) behavior or pathfinding code, try to keep them lightweight.

Placing Scenery here is just about a quick way to place obstacles to check movement mechanics are working correctly, later the majority (probably all for he prototype) of Scenery will be placed in from the start, ideally with a Map Editor if we have resources (time/energy).

So in short, it's OK to move placeScenery(). I think creating a MapEditor class is a good idea and that's an excellent place to move the placeScener() method. But I think you'd probably prefer that to be another PR so therefore it's probably best to leave placeScenery() here and handle it in a placeScenery() method.

File: TheGreatWar/src/TheGreatWar.cpp (hunk 9 above)
Comment (123samueld) [reply TheGreatWar/src/TheGreatWar.cpp:53]: Just to quickly reiterate, placeScenery() method (or similar) would be better in a new MapEditor class. However probably better that's in a new PR so please undo the movement & rename of placeScenery() and we'll cover that in it's own PR. I have no problem with you just doing that here but I think you'd prefer the new PR option.

**Review by 123samueld — CHANGES_REQUESTED** [review]
Please undo the placeScenery() method changes as described, that will be for another PR. Apart from that, good idea to split up the game and the engine! Thank you for doing that.

File: IsometricEngine/src/AgentManager.cpp (hunk 6 above)
Comment (danieldownes) [reply IsometricEngine/src/AgentManager.cpp:60]: Yes aware of the Terrain and Scenery differences. I think the destraction was its only called from TheGreatWar.cpp.

**Comment by danieldownes** [comment]
@123samueld placeScenery is restored and conflicts unconflicted.
